ethcore/src/client/traits.rs

```diff
@@ -48,12 +50,135 @@ use types::block_status::BlockStatus;
+			.expect("balance will return Some if given BlockId::Latest. balance was given BlockId::Latest \
+			Therefore balance has returned Some; qed")
+	}
+}
+
+/// Provides `chain_info` method
+pub trait ChainInfo {
+	/// Get blockchain information.
+	fn chain_info(&self) -> BlockChainInfo;
+}
 
+/// Provides verious information on a block by it's ID
+pub trait BlockInfo {
 	/// Get raw block header data by block id.
 	fn block_header(&self, id: BlockId) -> Option<encoded::Header>;
 
+	/// Get the best block header.
+	fn best_block_header(&self) -> encoded::Header;
+
+	/// Get raw block data by block header hash.
+	fn block(&self, id: BlockId) -> Option<encoded::Block>;
+}
+
+/// Provides `call_contract` method
+pub trait CallContract {
+	/// Like `call`, but with various defaults. Designed to be used for calling contracts.
+	fn call_contract(&self, id: BlockId, address: Address, data: Bytes) -> Result<Bytes, String>;
+}
+
+/// Provides various information on a transaction by it's ID
+pub trait TransactionInfo {
+	/// Get the hash of block that contains the transaction, if any.
+	fn transaction_block(&self, id: TransactionId) -> Option<H256>;
+}
+
+/// Provides information on a blockchain service and it's registry
+pub trait RegistryInfo {
+	/// Get the address of a particular blockchain service, if available.
+	fn registry_address(&self, name: String) -> Option<Address>;
+}
+
+// FIXME Why these methods belong to BlockChainClient and not MiningBlockChainClient?
+/// Provides methods to import block into blockchain
+pub trait ImportBlock {
+	/// Import a block into the blockchain.
+	fn import_block(&self, bytes: Bytes) -> Result<H256, BlockImportError>;
+
+	/// Import a block with transaction receipts. Does no sealing and transaction validation.
+	fn import_block_with_receipts(&self, block_bytes: Bytes, receipts_bytes: Bytes) -> Result<H256, BlockImportError>;
+}
+
+/// Provides methods to access chain state
+pub trait StateClient {
+	type State: StateInfo;
+
+	/// Get a copy of the best block's state.
+	fn latest_state(&self) -> Self::State;
+
+	/// Attempt to get a copy of a specific block's final state.
+	///
+	/// This will not fail if given BlockId::Latest.
+	/// Otherwise, this can fail (but may not) if the DB prunes state or the block
+	/// is unknown.
+	fn state_at(&self, id: BlockId) -> Option<Self::State>;
+}
+
+/// Provides `call` and `call_many` methods
+pub trait Call {
+	/// State type accepted by methods
+	type State: StateInfo;
+
+	/// Makes a non-persistent transaction call.
+	fn call(&self, tx: &SignedTransaction, analytics: CallAnalytics, state: &mut Self::State, header: &Header) -> Result<Executed, CallError>;
+
+	/// Makes multiple non-persistent but dependent transaction calls.
+	/// Returns a vector of successes or a failure if any of the transaction fails.
+	fn call_many(&self, txs: &[(SignedTransaction, CallAnalytics)], state: &mut Self::State, header: &Header) -> Result<Vec<Executed>, CallError>;
```

Comment:
why is `&mut Self::State` needed and not just `&Self::State`?

Reply:
Because it's accessed mutably in the client internals.

Reply:
This is the main point that requires mutable access: https://github.com/paritytech/parity/blob/client_split/ethcore/src/client/client.rs#L1216

Reply:
Isn't it contradictory with the method description (`non-persistent` calls)? Maybe we could add a comment that state changes should not be applied ?

Reply:
`non-persistent` = state is not applied
`dependent` = call `n+1` uses the state after `n` was applied

I'm fine with additional comment explaining that.

Reply:
Well, I don't see any problem, actuall. If you have a state that may be used for a call, you pass it "as is", but method prototype explicitly states, that passed state _may_ and probably _would_ be modified. If you're not comfortable with it, make a clone and pass it instead.

Reply:
> Isn't it contradictory with the method description (non-persistent calls)? 

I believe that's what `call_many` is for in the first place. It's ability to call several contracts on the same mutable state, instead of calling them one by one without having state shared between the calls.

Of course, we may think about more suitable API. Currently, it's just a way to solve the initial issue of client and miner decoupling.

Reply:
Probably `call` and `call_many` in trait definition could be unifined into `call(Vec<>)`

Reply:
yeah, the `call_many` functionality seems like a superset of `call`